storage/src/main/java/tech/pegasys/teku/storage/server/DepositStorage.java

```diff
@@ -44,22 +54,69 @@ public void stop() {
   }
 
   @Override
-  public SafeFuture<Stream<DepositsFromBlockEvent>> streamDepositFromBlockEvents() {
-    return SafeFuture.completedFuture(database.streamDepositsFromBlocks());
+  public SafeFuture<ReplayDepositsResult> replayDepositEvents() {
+    return replayResult.get();
   }
 
-  @Override
-  public SafeFuture<Optional<MinGenesisTimeBlockEvent>> getMinGenesisTimeBlockEvent() {
-    return SafeFuture.completedFuture(database.getMinGenesisTimeBlock());
+  private ReplayDepositsResult replayDeposits() {
+    isSyncingFromDatabase = true;
+    final DepositSequencer depositSequencer =
+        new DepositSequencer(eth1EventsChannel, database.getMinGenesisTimeBlock());
+    try (Stream<DepositsFromBlockEvent> eventStream = database.streamDepositsFromBlocks()) {
+      eventStream.forEach(depositSequencer::depositEvent);
+    }
+    depositSequencer.depositsComplete();
+    isSyncingFromDatabase = false;
+    return depositSequencer.done();
   }
 
   @Override
   public void onDepositsFromBlock(final DepositsFromBlockEvent event) {
-    database.addDepositsFromBlockEvent(event);
+    if (!isSyncingFromDatabase) {
+      database.addDepositsFromBlockEvent(event);
+    }
   }
 
   @Override
   public void onMinGenesisTimeBlock(final MinGenesisTimeBlockEvent event) {
-    database.addMinGenesisTimeBlock(event);
+    if (!isSyncingFromDatabase) {
+      database.addMinGenesisTimeBlock(event);
+    }
+  }
+
+  static class DepositSequencer {
+    private final Eth1EventsChannel eth1EventsChannel;
+    private final Optional<MinGenesisTimeBlockEvent> genesis;
+    private boolean isGenesisDone;
+    private UnsignedLong lastDeposit;
+
+    public DepositSequencer(
+        Eth1EventsChannel eventChannel, Optional<MinGenesisTimeBlockEvent> genesis) {
+      this.eth1EventsChannel = eventChannel;
+      this.genesis = genesis;
+      this.isGenesisDone = false;
+    }
+
+    public void depositEvent(final DepositsFromBlockEvent event) {
+      if (genesis.isPresent()
+          && !isGenesisDone
+          && genesis.get().getBlockNumber().compareTo(event.getBlockNumber()) < 0) {
```

Comment:
I think this should be less than or equal to:
```suggestion
          && genesis.get().getBlockNumber().compareTo(event.getBlockNumber()) <= 0) {
```

Reply:
And also this line should be before posting the MinGenesisTimeBlockEvent to eth1EventsChannel :
```
eth1EventsChannel.onDepositsFromBlock(event);
```

Reply:
Otherwise, we would miss deposits, if there were any in the MinGenesisTime Eth1Block.